components/async-container/async-container-mixin.js

```diff
@@ -0,0 +1,78 @@
+
+export const asyncStates = {
+	initial: 'initial',
+	pending: 'pending',
+	success: 'success',
+	failure: 'failure'
+};
+
+export const AsyncContainerMixin = superclass => class extends superclass {
+
+	static get properties() {
+		return {
+			asyncState: { type: String }
+		};
+	}
+
+	constructor() {
+		super();
+		this._initializeAsyncState();
+		if (!this.asyncContainer) {
+			this.addEventListener('pending-state', this._handleAsyncItemState.bind(this));
+		}
+	}
+
+	firstUpdated() {
+		if (!this.asyncContainer) return;
+		const container = this.asyncContainer();
+		if (container) {
+			container.addEventListener('pending-state', this._handleAsyncItemState.bind(this));
+		}
+	}
+
+	resetAsyncState() {
+		this._initializeAsyncState();
+	}
+
+	async _handleAsyncItemState(e) {
+		const promise = e.detail.promise;
+
+		if (!promise) return;
+		if (this.asyncState === asyncStates.success
+			|| this.asyncState === asyncStates.failure) return;
```

Comment:
This bit is interesting to me. I'm assuming this gets hit if some async children register, then go into success or failure state, then a new child registers?

I guess the question is: do we expect folks to always have to reset the state before new things register, or does it make sense for it to be able to go from initial -> pending -> success || failure -> pending -> success || failure? Essential is it a one-way workflow or cyclical?

Reply:
Or a related question: the code is assuming that all the registrations happen before the first promise resolves. Hopefully that's true?

Reply:
> This bit is interesting to me. I'm assuming this gets hit if some async children register, then go into success or failure state, then a new child registers?

That's correct.  The `success` and `failure` are the two possible end-states.  I am starting to wonder though, whether it should even have a failure state... it's only useful if the consumer is going to use it, and if we do keep it, then it forces consumers to check for either even if they don't care, unless we provide some `hasPending` sugar.

> Essential is it a one-way workflow or cyclical?

Good question.  Consider use-case where a list loads, and then the user clicks load-more, which may dispatch a new `pending-state` event. I'm not sure if it would dispatch the event in such a case, but the result could be several more API requests for nested async components, and they might dispatch that even too.  In such case, I don't think we'd want the initial list to be replaced while the load-more operation is in progress.

The async components themselves don't know their context, and whether to dispatch such events.  And, the async container mixin doesn't really know which async operations to care about either. Currently it's just assuming the first series of async events are the ones to watch, and thereafter the consumer would tell it to reset.

> Or a related question: the code is assuming that all the registrations happen before the first promise resolves. Hopefully that's true?

Yes that's absolutely correct. It's assuming the requests would be overlapping.  I am a bit concerned about this.  We could build some buffer into it, but this seems flakey.

Reply:
I like those responses.

The use case of load-more is a good one to walk through, and I agree with your conclusion that only the developer in charge of the whole flow knows how things should behave -- so requiring them to reset things after the first load if they want the workflow to start over seems reasonable.

Not having a failure state is also interesting. I think I'd be on board with abandoning it, since if someone wanted to handle it they could always do so manually -- after all they're the ones rejecting promises.